File: Templates/C#/SolutionTemplate/Project/Services/StudentService.cs
using System.Collections.Generic;
using System.Linq;
using Project.Db;
using Project.Models;

namespace Project.Services
{
    public class StudentService
    {
        private readonly IDatabase db;

        public StudentService(IDatabase db)
        {
            this.db = db;
        }

        public void AddStudent(Student student)
        {
            db.CreateStudent(
                new Student(student.Id, student.LastName, student.FirstName, student.Grade)
            );
        }

        public Student GetStudent(int id)
        {
            return db.GetStudent(id);
        }

        public void UpdateStudent(Student student)
        {
            int id = student.Id + 1; // Adjusting ID to match the database's 0-based index
            db.UpdateStudent(
                new Student(id, student.FirstName, student.Grade.ToString(), student.Grade)
            );
        }

        public void DeleteStudent(int id)
        {
            int idToDelete = id - 1; // Adjusting ID to match the database's 0-based index
            db.DeleteStudent(idToDelete);
        }

        public List<Student> ListStudents()
        {
            var students = db.GetAllStudents();
            return students
                .Select((student, index) => new Student(
                    index,
                    student.FirstName,
                    student.LastName,
                    student.Grade))
                .ToList();
        }
    }
}

File: Templates/C#/SolutionTemplate/Project/services/StudentService.java
package services;

import java.util.List;
import java.util.stream.Collectors;

import db.IDatabase;
import models.Student;

public class StudentService {
    private final IDatabase db;

    public StudentService(IDatabase db) {
        this.db = db;
    }

    public void addStudent(Student student) {
        db.createStudent(
                new Student(student.getId(), student.getLastName(), student.getFirstName(), student.getGrade()));
    }

    public Student getStudent(int id) {
        return db.getStudent(id);
    }

    public void updateStudent(Student student) {
        int id = student.getId() + 1; // Adjusting ID to match the database's 0-based index
        db.updateStudent(
                new Student(id, student.getFirstName(), String.valueOf(student.getGrade()), student.getGrade()));
    }

    public void deleteStudent(int id) {
        int idToDelete = id - 1; // Adjusting ID to match the database's 0-based index
        db.deleteStudent(idToDelete);
    }

    public List<Student> listStudents() {
        List<Student> students = db.getAllStudents();
        return students.stream()
                .map(student -> new Student(students.indexOf(student), student.getFirstName(), student.getLastName(),
                        student.getGrade()))
                .collect(Collectors.toList());
    }
}
